pkg/trace/metrics/timing/timing.go

```diff
@@ -0,0 +1,143 @@
+// Package timing is used to aggregate timing calls within hotpaths to avoid using
+// repeated statsd calls. The package has a default set that reports at 10 second
+// intervals and can be used directly. If a different behaviour or reporting pattern
+// is desired, a custom Set may be created.
+package timing
+
+import (
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/trace/atomic"
+	"github.com/DataDog/datadog-agent/pkg/trace/metrics"
+)
+
+// AutoreportInterval specifies the interval at which the default set reports.
+const AutoreportInterval = 10 * time.Second
+
+var (
+	defaultSet = NewSet()
+	stopReport = defaultSet.Autoreport(AutoreportInterval)
+)
+
+// Since records the duration for the given metric name as time passed since start.
+// It uses the default set which is reported at 10 second intervals.
+func Since(name string, start time.Time) { defaultSet.Since(name, start) }
+
+// Flush reports the default set. It can be useful to call when the program
+// exits prematurely.
+func Flush() { stopReport() }
+
+// Set represents a set of metrics that can be used for timing. Use NewSet
+// to create one and Report (or Autoreport) to submit metrics. It is safe for
+// concurrent use.
+type Set struct {
```

Comment:
Not sure what go standards are like, but calling this such generic name like `Set` could be confusing. Just reading this `NewSet(...) *Set` I'd assume it's a set data structure in the literal sense. Maybe a better name would be `MetricCounters` or `MetricSet` or something.

Reply:
Generally, in Go, the best way to think about an identifier's name is within the context of the package that it is part of. In this case, think of this structure as `timing.Set`, not `Set` on its own. This is so because it will be used as such. I completely understand what you're saying but I think that at the same time saying that this is a set of timers is fine.

A better name would perhaps be `timing.Metrics` which would be more accurate. But I think the plan (in my mind) is to eventually move all of this into the `metrics` package as potentially `metrics.Set` (a set of metrics) and apply the same logic for other metrics, such as counters, gauges, etc. This will help us remove a lot of extra logic and counters which are currently all over the place in packages such as `info`.

So I'd say it's fine to leave it as is for now.